File: ps/leetcode/ValidParentheses.cpp
#include <iostream>
#include <stack>
using namespace std;
class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        int slen = s.size();
        for(int i=0; i<slen; ++i){
            char c = s[i];
            if(c=='(' || c=='[' || c=='{'){
                stk.push(c);
            }
            else {
                if(stk.empty()) return false;
                if(c==')' && stk.top()=='(') stk.pop();
                else if(c==']' && stk.top()=='[') stk.pop();
                else if(c=='}' && stk.top()=='{') stk.pop();
                else return false;
            }
        }
        return stk.empty();
    }
};

File: ps/leetcode_java/src/ValidParentheses/ValidParentheses.java
package ValidParentheses;

import java.util.Stack;

public class ValidParentheses {
    final char[] open = {'(', '[', '{'};
    final char[] close = {')', ']', '}'};
    public int getOpenIndex(char ch){
        for(int i=0; i<3; ++i){
            if(ch == open[i])
                return i;
        }
        return -1;
    }
    public int getCloseIndex(char ch){
        for(int i=0; i<3; ++i){
            if(ch == close[i])
                return i;
        }
        return -1;
    }

    public boolean isValid(String s) {
        Stack<Integer> stk = new Stack<>();
        int len = s.length();
        for(int i=0; i<len; ++i){
            char ch = s.charAt(i);
            int oi = getOpenIndex(ch);
            if(oi!=-1) stk.push(oi);
            else{
                int ci = getCloseIndex(ch);
                if(stk.empty() || stk.peek() != ci) return false;
                stk.pop();
            }
        }
        return stk.empty();
    }
}
